File: examples/python/option_info.py
###
# Bitwuzla: Satisfiability Modulo Theories (SMT) solver.
#
# Copyright (C) 2023 by the authors listed in the AUTHORS file at
# https:#github.com/bitwuzla/bitwuzla/blob/main/AUTHORS
#
# This file is part of Bitwuzla under the MIT license. See COPYING for more
# information at https:#github.com/bitwuzla/bitwuzla/blob/main/COPYING
##

import pytest

from bitwuzla import *

if __name__ == '__main__':

    # First, create a Bitwuzla options instance.
    options = Options()

    # Set some options to illustrate current vs default value.
    options.set(Option.PRODUCE_MODELS, True)
    options.set(Option.VERBOSITY, 2)
    options.set(Option.BV_SOLVER, 'prop')

    # Then iterate over all available configuration options and extract info.
    for opt in Option:
        info = OptionInfo(options, opt)
        print(f'  long name:   {info.lng()}')
        print(f'  short name:  {info.shrt() if info.shrt() else "-"}')
        print('  kind:        ', end = '')
        if info.kind() == OptionInfoKind.BOOL:
            print('bool')
            print('  values:')
            print(f'  + current:   {info.cur()}')
            print(f'  + default:   {info.dflt()}')
        elif info.kind() == OptionInfoKind.NUMERIC:
            print('numeric')
            print('  values:')
            print(f'  + current:   {info.cur()}')
            print(f'  + default:   {info.dflt()}')
            print(f'  + min:       {info.min()}')
            print(f'  + max:       {info.max()}')
        else:
            print('modes')
            print('  values:')
            print(f'  + current:   {info.cur()}')
            print(f'  + default:   {info.dflt()}')
            print(f'  + modes:     {info.modes()}')
        print(f'  description: {info.description()}\n')

File: examples/c/option_info.c
/***
 * Bitwuzla: Satisfiability Modulo Theories (SMT) solver.
 *
 * Copyright (C) 2023 by the authors listed in the AUTHORS file at
 * https://github.com/bitwuzla/bitwuzla/blob/main/AUTHORS
 *
 * This file is part of Bitwuzla under the MIT license. See COPYING for more
 * information at https://github.com/bitwuzla/bitwuzla/blob/main/COPYING
 */

#include <bitwuzla/c/bitwuzla.h>
#include <stdio.h>

int
main()
{
  // First, create a Bitwuzla options instance.
  BitwuzlaOptions* options = bitwuzla_options_new();
  // Set some options to illustrate current vs default value.
  bitwuzla_set_option(options, BITWUZLA_OPT_PRODUCE_MODELS, 1);
  bitwuzla_set_option(options, BITWUZLA_OPT_VERBOSITY, 2);
  bitwuzla_set_option_mode(options, BITWUZLA_OPT_BV_SOLVER, "prop");

  // Then iterate over all available configuration options and extract info.
  BitwuzlaOptionInfo info;
  for (uint32_t i = 0; i < BITWUZLA_OPT_NUM_OPTS; ++i)
  {
    BitwuzlaOption opt = (BitwuzlaOption) i;
    bitwuzla_get_option_info(options, opt, &info);
    printf("- long name:   %s\n", info.lng);
    printf("  short name:  %s\n", (info.shrt ? info.shrt : "-"));
    printf("  kind:        ");
    if (info.is_numeric)
    {
      printf("numeric\n");
      printf("  values:\n");
      printf("  + current:   %lu\n", info.numeric.cur);
      printf("  + default:   %lu\n", info.numeric.dflt);
      printf("  + min:       %lu\n", info.numeric.min);
      printf("  + max:       %lu\n", info.numeric.max);
    }
    else
    {
      printf("modes\n");
      printf("  values:\n");
      printf("  + current:   %s\n", info.mode.cur);
      printf("  + default:   %s\n", info.mode.dflt);
      printf("  + modes:     {");
      for (size_t i = 0, n = info.mode.num_modes; i < n; ++i)
      {
        printf("%s %s", (i > 0 ? "," : ""), info.mode.modes[i]);
      }
      printf(" }\n");
    }
    printf("  description: %s\n\n", info.description);
  }
}
